src/template.cpp

```diff
@@ -0,0 +1,146 @@
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// Positive substring: return a substring of `str``going from `start` to `end`;
+// if `end` is earlier than `start`, return empty string.
+std::string pos_substr(std::string str, int start, int end) {
+  if (end < start)
+    return "";
+  else
+    return str.substr(start, end - start + 1);
+}
+
+// Break template text into character vector. The first element element of the
+// resulting vector is HTML, the next is R code, and they continue alternating.
+// [[Rcpp::export]]
+CharacterVector template_dfa(CharacterVector x) {
+  enum State {
+    html,
+    code,
+    html_oneOpenBracket,
+    code_oneCloseBracket,
+    code_string1,
+    code_string1_backslash,
+    code_string2,
+    code_string2_backslash,
+    code_backtick,
+    code_backtick_backslash,
+    code_backslash
+  };
+
+  if (x.length() != 1) {
+    stop("Input HTML must be a character vector of length 1");
+  }
+  std::string input = Rcpp::as<std::string>(x[0]);
+  CharacterVector pieces(0);
+
+  int i = 0;
+  int pieceStartIdx = 0;
+  int len = input.length();
+  char c;
+  State state = html;
+  while (i < len) {
```

Comment:
You don't have to change this, but, I tend to use a for-loop instead of while-loop, to make it harder to accidentally bypass `i++` (for example, a `continue` statement anywhere below--you don't have any but it can sometimes be useful).